// react-water-wave/lib/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var PropTypes = require("prop-types");
var clean_react_props_1 = require("clean-react-props");
var $ = require("jquery");
require("jquery.ripples");
var useRipples = function (_a) {
    var imageUrl = _a.imageUrl, dropRadius = _a.dropRadius, perturbance = _a.perturbance, resolution = _a.resolution, interactive = _a.interactive, crossOrigin = _a.crossOrigin, rippleRef = _a.rippleRef;
    var target = React.useRef({
        ripples: function (_arg) {
            var _args = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                _args[_i - 1] = arguments[_i];
            }
        },
    });
    React.useEffect(function () {
        target.current = $(rippleRef.current);
        target.current.ripples({
            imageUrl: imageUrl,
            dropRadius: dropRadius,
            perturbance: perturbance,
            resolution: resolution,
            interactive: interactive,
            crossOrigin: crossOrigin,
        });
        return function () { return target.current.ripples('destroy'); };
    }, []);
    var destroy = function () {
        target.current.ripples('destroy');
    };
    var drop = function (_a) {
        var _b = _a === void 0 ? {
            x: undefined,
            y: undefined,
            radius: undefined,
            strength: undefined,
        } : _a, _c = _b.x, x = _c === void 0 ? undefined : _c, _d = _b.y, y = _d === void 0 ? undefined : _d, _e = _b.radius, radius = _e === void 0 ? undefined : _e, _f = _b.strength, strength = _f === void 0 ? undefined : _f;
        target.current.ripples('drop', x, y, radius, strength);
    };
    var pause = function () {
        target.current.ripples('pause');
    };
    var play = function () {
        target.current.ripples('play');
    };
    var hide = function () {
        target.current.ripples('hide');
    };
    var show = function () {
        target.current.ripples('show');
    };
    var set = function (_a) {
        var _b = _a === void 0 ? {
            property: undefined,
            value: undefined,
        } : _a, _c = _b.property, property = _c === void 0 ? undefined : _c, _d = _b.value, value = _d === void 0 ? undefined : _d;
        target.current.ripples('set', property, value);
    };
    var updateSize = function () {
        target.current.ripples('updateSize');
    };
    return {
        destroy: destroy,
        pause: pause,
        play: play,
        hide: hide,
        show: show,
        drop: drop,
        set: set,
        updateSize: updateSize,
    };
};
var WaterEffect = function (_a) {
    var imageUrl = _a.imageUrl, dropRadius = _a.dropRadius, perturbance = _a.perturbance, resolution = _a.resolution, interactive = _a.interactive, crossOrigin = _a.crossOrigin, children = _a.children, otherProps = __rest(_a, ["imageUrl", "dropRadius", "perturbance", "resolution", "interactive", "crossOrigin", "children"]);
    var rippleRef = React.useRef(null);
    var _b = useRipples({
        imageUrl: imageUrl,
        dropRadius: dropRadius,
        perturbance: perturbance,
        resolution: resolution,
        interactive: interactive,
        crossOrigin: crossOrigin,
        rippleRef: rippleRef,
    }), destroy = _b.destroy, pause = _b.pause, play = _b.play, hide = _b.hide, show = _b.show, drop = _b.drop, set = _b.set, updateSize = _b.updateSize;
    return (React.createElement("div", __assign({ ref: rippleRef }, clean_react_props_1.default(otherProps)), children({
        destroy: destroy,
        pause: pause,
        play: play,
        hide: hide,
        show: show,
        drop: drop,
        set: set,
        updateSize: updateSize,
    })));
};
WaterEffect.propTypes = {
    imageUrl: PropTypes.string,
    dropRadius: PropTypes.number,
    perturbance: PropTypes.number,
    resolution: PropTypes.number,
    interactive: PropTypes.bool,
    crossOrigin: PropTypes.string,
    children: PropTypes.func,
};
WaterEffect.defaultProps = {
    imageUrl: '',
    dropRadius: 20,
    perturbance: 0.03,
    resolution: 256,
    interactive: true,
    crossOrigin: '',
    children: function () { },
};
exports.default = WaterEffect;
